### PR TITLE
feat(router): introduce `ParamMap` to access parameters

### PR DESCRIPTION
**Pull request by vicb**

### CODE REVIEW

File: packages/router/src/router_state.ts [hunk 1]
```diff
@@ -149,6 +155,21 @@ export class ActivatedRoute {
   /** The path from the root of the router state tree to this route */
   get pathFromRoot(): ActivatedRoute[] { return this._routerState.pathFromRoot(this); }
 
+  get paramMap(): Observable<ParamMap> {
```
Comment (DzmitryShylovich) [inline packages/router/src/router_state.ts]: `params` will be deprecated?

File: packages/router/src/shared.ts [hunk 2]
```diff
@@ -27,6 +27,62 @@ export type Params = {
...
+  get(name: string): string;
+  /**
+   * Return an array of values for the given parameter name.
+   *
+   * If there is no such parameter, an empty array is returned.
+   */
+  getAll(name: string): string[];
+}
+
+class ParamsAsMap implements ParamMap {
```
Comment (tbosch) [inline packages/router/src/shared.ts]: Add a comment: See UrlSearchParam

File: packages/router/src/shared.ts [hunk 3]
```diff
@@ -27,6 +27,62 @@ export type Params = {
   [key: string]: any
 };
 
+/**
```
Comment (tbosch) [inline packages/router/src/shared.ts]: Maybe create a bug for angular.io to update the guide and add to commit message...

File: packages/router/src/shared.ts (hunk 2 above)
Comment (tbosch) [inline packages/router/src/shared.ts]: Maybe add a true unit test for this..

**Review by tbosch — APPROVED** [review]

File: packages/router/src/router_state.ts (hunk 1 above)
Comment (vicb) [inline packages/router/src/router_state.ts]: not for now

**Review by mhevery — APPROVED** [review]

**Pull request merged**